File: Java_Dsc/InsertionSort.java
import java.util.*;

class InsertionSort {

    public static int[] doInsertionSort(int[] input, boolean reverse){
        int temp; // Constant O(1) memory
        for(int i = 1; i< input.length; i++){
            for(int j=i; j>0; j--){                 // O(n^2) time
                if(reverse == false) {
                    if(input[j] < input[j-1]){
                        temp = input[j];
                        input[j] = input[j-1];
                        input[j-1] = temp;
                    }
                } else {
                    if(input[j] > input[j-1]){
                        temp = input[j];
                        input[j] = input[j-1];
                        input[j-1] = temp;
                    }
                }
            }
        }
        return input;
    }

    public static int[] doRecursiveInsertionSort(int[] array, int n){
        int i;
        if (n > 1) {
            doRecursiveInsertionSort(array, n - 1);
            int last = array[n-1];
            i = n - 2;
            while (i >= 0  &&  array[i] > last) {
                array[i + 1] = array[i];
                i = i - 1;
            }
            array[i + 1] = last;
        }
        return array;

    }

    public static void main(String[] args){
        int[] arr1 = {10,34,2,56,7,67,88,42};
        System.out.println("Input array is: " + Arrays.toString(arr1));

        int[] arr2 = doInsertionSort(arr1, false);
        System.out.println("Insertion sort is: " + Arrays.toString(arr2));
        
        int[] arr3 = doInsertionSort(arr1, true);
        System.out.println("Reverse Insertion sort is: " + Arrays.toString(arr3));

        int[] arr4 = doRecursiveInsertionSort(arr1, arr1.length);
        System.out.println("Recursive Insertion sort is: " + Arrays.toString(arr4));
    }
}

/* Output:
 * Input array is: [10, 34, 2, 56, 7, 67, 88, 42]
 * Insertion sort is: [2, 7, 10, 34, 42, 56, 67, 88]
 * Reverse Insertion sort is: [88, 67, 56, 42, 34, 10, 7, 2]
 * Recursive Insertion sort is: [2, 7, 10, 34, 42, 56, 67, 88]
 */

File: Python_Dsc/InsertionSort.py
# Python program for implementation of Insertion Sort

# Function to do insertion sort iteratively
def iterativeInsertionSort(arr,reverse):
    #Traverse through l to len(arr)
    for i in range(1,len(arr)):
        for j in range(i, 0, -1):
            if reverse == False: 
                if(arr[j] < arr[j-1]):
                    temp = arr[j]
                    arr[j] = arr[j-1]
                    arr[j-1] = temp
            else:
                if(arr[j] > arr[j-1]):
                    temp = arr[j]                
                    arr[j] = arr[j-1]                
                    arr[j-1] = temp
    return arr      

# Recursive Python program for insertion sort
def recursiveInsertionSort(arr,n):
    #base case
    if n<=1:
        return
    
    #sort first n-1 elements
    recursiveInsertionSort(arr,n-1)

    '''Insert last element at its correct position
            in sorted array.'''
    last = arr[n-1]
    j = n-2
    
    # Move elements of arr[0..i-1], that are
    # greater than key, to one position ahead
    # of their current position 

    while ( j >=0 and arr[j] > last):
        arr[j+1] = arr[j];
        j = j-1
    
    arr[j+1] = last
    return arr

# Driver code to test above
arr = [12, 11, 13, 5, 6]
print("Input Array is:")
print arr
iterativeInsertionSort(arr, False)
print ("Iterated Insertion : Sorted array is:")
print arr

arr1 = [15, 13, 16, 5, 6]
print("Input Array is:")
print arr1
iterativeInsertionSort(arr1, True)
print ("Reverse Iterated Insertion : Sorted array is:")
print arr1


arr2 = [15, 12, 16, 5, 7]
print("Input Array is:")
print arr2
n = len(arr2)
recursiveInsertionSort(arr2, n)
print ("Recursive Insertion : Sorted array is:")
print arr2

''' Output:
    Input Array is:
    [12, 11, 13, 5, 6]
    Iterated Insertion : Sorted array is:
    [5, 6, 11, 12, 13]
    Input Array is:
    [15, 13, 16, 5, 6]
    Reverse Iterated Insertion : Sorted array is:
    [16, 15, 13, 6, 5]
    Input Array is:
    [15, 12, 16, 5, 7]
    Recursive Insertion : Sorted array is:
    [5, 7, 12, 15, 16]
'''
